sdk/core/core/test/cmocka/test_az_http.c

```diff
@@ -512,6 +521,60 @@ static void test_http_response_header_validation_space(void** state)
   }
 }
 
+static void test_http_response_append(void** state)
+{
+  (void)state;
+  {
+    uint8_t buffer[10];
+    az_http_response response = { 0 };
+    az_span append_this = AZ_SPAN_FROM_STR("0123456789");
+    assert_return_code(az_http_response_init(&response, AZ_SPAN_FROM_BUFFER(buffer)), AZ_OK);
+    assert_return_code(az_http_response_append(&response, append_this), AZ_OK);
+    assert_memory_equal(buffer, az_span_ptr(append_this), 10);
+  }
+}
+
+static void test_http_response_append_overflow(void** state)
+{
+  (void)state;
+  {
+    uint8_t buffer[10] = "..........";
+    az_http_response response = { 0 };
+    az_span append_this = AZ_SPAN_FROM_STR("0123456789123456");
+
+    assert_return_code(az_http_response_init(&response, AZ_SPAN_FROM_BUFFER(buffer)), AZ_OK);
+
+    int32_t response_written_state = response._internal.written;
+
+    az_result result = az_http_response_append(&response, append_this);
+    assert_true(result == AZ_ERROR_INSUFFICIENT_SPAN_SIZE);
+
+    // make sure buffer content didn't change
+    assert_memory_equal(buffer, "..........", 10);
+
+    // make sure response state didn't change
+    assert_int_equal(response_written_state, response._internal.written);
```

Comment:
nit: I would remove the access to `_internal` within the new test being added. It isn't necessary and we should try to only test the public surface area directly to avoid tight coupling and leaking implementation detail. It becomes a bit annoying to have to update tests when we change the internal state of things.

The other way to test this behavior would be to check the output buffer, which you already do, but then, continue to use the `response` object after.

So, I recommend merging `test_http_response_append_overflow` and `test_http_response_append_overflow_on_second_call`, and do the `az_http_response_append(&response, this_will_fit_once)` twice right after `az_http_response_append(&response, append_this)`. If the internal state was broken, the subsequent call would fail somehow.

That is the only thing that is observable to the end user anyway, so let's only test that.

Reply:
use of internal is all over the place.
I think it is fine for now. We can think about if we want to avoid in testing or not.
I believe is not so bad.